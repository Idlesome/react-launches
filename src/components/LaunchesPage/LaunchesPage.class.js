/**
 * Container, api manager and layout for displaying a list of SpaceX launches
 */

import React from "react";
import PropTypes from "prop-types";

import axios from "axios";

import { withStyles } from "@material-ui/core/styles";
import Grid from "@material-ui/core/Grid";
import Typography from "@material-ui/core/Typography";
import Button from "@material-ui/core/Button";

import TextFilter from "components/TextFilter";
import LaunchCard from "components/LaunchCard/";
import Loading from "components/Loading";

const SPACEWATCH_LAUNCHES_API = "https://api.spacexdata.com/v3/launches";

const styles = {
  layout: {
    width: "auto",
    marginLeft: "2vmin",
    marginRight: "2vmin"
  },
  button: {
    margin: "5px"
  }
};

// Write tests, prop types, and document prop types
// and components
class LaunchesPage extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      loading: true,
      launches: [],
      error: null,
      textFilter: "",
      offset: 0
    };

    // Limit launches loaded to 6
    // This will also indicate "Load More" button to load 6 more launches
    this.limit = 6;
  }

  componentDidMount() {
    this.loadLaunches();
  }

  loadLaunches = (params = {}) => {
    this.setState(
      {
        loading: true
      },
      () => {
        axios
          .get(SPACEWATCH_LAUNCHES_API, {
            params: { limit: this.limit, ...params }
          })
          .then(result => {
            this.setState({
              // Append new data onto current data
              launches: this.state.launches.concat(result.data),
              loading: false,
              // Increase offset
              offset: this.state.offset + this.limit
            });
          })
          .catch(error =>
            this.setState({
              error,
              loading: false
            })
          );
      }
    );
  };

  loadMore = () => {
    this.loadLaunches({ offset: this.state.offset + this.limit });
  };

  /** Filter launches based on available filters */
  launchFilter = launch => {
    const { textFilter } = this.state;

    let { details, mission_name, launch_year } = launch;

    // Default to empty string if null
    details = details || "";

    return (
      mission_name.toLowerCase().includes(textFilter) ||
      launch_year.includes(textFilter) ||
      details.toLowerCase().includes(textFilter)
    );
  };

  onFilterChange = event => {
    this.setState({
      textFilter: event.target.value
    });
  };

  render() {
    const { loading, launches, error } = this.state;
    const { classes } = this.props;

    if (error) {
      return <Typography component="p">{error.message}</Typography>;
    }

    return (
      <div className={classes.layout}>
        <Grid container spacing={4}>
          <Grid item xs={12}>
            {loading && <Loading />}
            <TextFilter
              label="Filter launches"
              onChange={this.onFilterChange}
            />
          </Grid>

          {launches.filter(this.launchFilter).map(launch => (
            <LaunchCard {...launch} key={launch.flight_number} />
          ))}

          <Grid item xs={12}>
            <Button
              disabled={loading}
              variant="outlined"
              className={classes.button}
              onClick={this.loadMore}>
              Load More
            </Button>
          </Grid>
        </Grid>
      </div>
    );
  }
}

LaunchesPage.propTypes = {
  classes: PropTypes.object.isRequired
};

export default withStyles(styles)(LaunchesPage);
