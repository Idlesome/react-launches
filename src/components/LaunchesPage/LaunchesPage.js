/**
 * Saga style Launches Page component which uses redux to dispatch
 * the api request
 *
 */

import React, { useEffect, useState } from "react";
import PropTypes from "prop-types";

import { withStyles } from "@material-ui/core/styles";
import Grid from "@material-ui/core/Grid";
import Typography from "@material-ui/core/Typography";
import Button from "@material-ui/core/Button";

import TextFilter from "components/TextFilter";
import LaunchCard from "components/LaunchCard/";
import Loading from "components/Loading";

const styles = {
  layout: {
    width: "auto",
    marginLeft: "2vmin",
    marginRight: "2vmin"
  },
  button: {
    margin: "5px"
  }
};

function LaunchesPage(props) {
  const {
    launches,
    dispatchLoadLaunches,
    error,
    loading,
    // Limit launches loaded to default of 6
    // This will also indicate "Load More" button to load 6 more launches
    limit = 6
  } = props;

  const [textFilter, setTextFilter] = useState("");
  const [offset, setOffset] = useState(0);

  // Load launches on mount
  useEffect(() => {
    loadLaunches();
    // Explictly ensuring this only gets run once
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  /**
   * Loads launches. After first call it loads 6 more
   *
   * @param {Object} params optionally override params offset and limit
   */
  const loadLaunches = (params = {}) => {
    // Increase offset
    setOffset(offset + limit);
    // Dispatch load event
    dispatchLoadLaunches({ params: { offset, limit, ...params } });
  };

  /** Load More button event handler */
  const onLoadMore = () => {
    loadLaunches();
  };

  /** Filter launches based on available filters */
  const launchFilter = launch => {
    let { details, mission_name, launch_year } = launch;

    // Default to empty string if null
    details = details || "";

    return (
      mission_name.toLowerCase().includes(textFilter) ||
      launch_year.includes(textFilter) ||
      details.toLowerCase().includes(textFilter)
    );
  };

  const onFilterChange = event => {
    setTextFilter(event.target.value);
  };

  const { classes } = props;

  if (error) {
    return <Typography component="p">{error.message}</Typography>;
  }

  return (
    <div className={classes.layout}>
      <Grid container spacing={4}>
        <Grid item xs={12}>
          {loading && <Loading />}
          <TextFilter
            label="Filter launches"
            onChange={e => onFilterChange(e)}
          />
        </Grid>

        {launches.filter(launchFilter).map(launch => (
          <LaunchCard {...launch} key={launch.flight_number} />
        ))}

        <Grid item xs={12}>
          <Button
            disabled={loading}
            variant="outlined"
            className={classes.button}
            onClick={onLoadMore}>
            Load More
          </Button>
        </Grid>
      </Grid>
    </div>
  );
}

LaunchesPage.propTypes = {
  classes: PropTypes.object.isRequired
};

export default withStyles(styles)(LaunchesPage);
