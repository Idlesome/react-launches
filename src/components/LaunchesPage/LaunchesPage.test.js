import React from "react";

import { createMount } from "@material-ui/core/test-utils";

import Enzyme from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import LaunchesPage from "./LaunchesPage";

Enzyme.configure({ adapter: new Adapter() });

describe("<LaunchesPage />", () => {
  let mount;

  const mockLoadLaunches = jest.fn();

  beforeAll(() => {
    mount = createMount();
  });

  afterAll(() => {
    mount.cleanUp();
  });

  it("attempts to load launches on mount", done => {
    const wrapper = mount(
      <LaunchesPage dispatchLoadLaunches={mockLoadLaunches} launches={[]} />
    );

    wrapper.update();

    expect(mockLoadLaunches.mock.calls.length).toBe(1);

    done();
  });

  it("should render component", () => {
    const wrapper = mount(
      <LaunchesPage dispatchLoadLaunches={mockLoadLaunches} launches={[]} />
    );
    expect(wrapper.find("div")).toMatchSnapshot();
  });
});
