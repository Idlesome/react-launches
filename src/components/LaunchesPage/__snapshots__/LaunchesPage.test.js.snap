// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LaunchesPage /> should render component 1`] = `ReactWrapper {}`;
